KD-4000: fix for KD-543 / 8f1b7b8 after which only half of the items became added,

nugged opened the pull request:
The i++ being in the for clause is sufficient because the $barcodeobj->next_value and C4::Barcodes::ValueBuilder::hbyyyyincr::get_barcode statements return a valid barcode always except when we are out of barcodes. 

Therefore, we can remove the other i++ statement inside the for a loop. In the case where we run out of barcodes, this adds the error "no_next_barcode" to the list of errors.

––-

The previous discussion was in PR https://github.com/KohaSuomi/Koha/pull/113
(also to refer related commits from IRC discussion: https://github.com/KohaSuomi/Koha/commit/8f1b7b88848fd9b0fc767e3ff7dcdc7a56f6ebff, https://github.com/KohaSuomi/kohasuomi/commit/ada6401c2c1e29bf4bb1a33e0153aa041ebac196 )

kohaputti commented:
Please make sure the barcode retrieval succeeds always when possible. The only time it cannot succeed is when the barcode space is totally consumed. It would be nice to have a error message in that case but the current code doesn't do that so I would leave showing an error message to another pull request.

Also, you can just have one pull request so the progress is easy to follow.

nugged commented:
I insist we shouldn't keep `$i++` all the time, but only when copy added. 
So I will rework the code back to count only really added + prevent border condition of the infinite loop like I did first time.

The reason, for ex, lines 664 and 665 uses the counter as first-sign of successful adding:
```
$barcodevalue = $new_barcode if ($barcodePreference eq 'hbyyyyincr' && ($i > 0 || $exist_itemnumber));
$barcodevalue = $barcodeobj->next_value($oldbarcode) if ($barcodePreference ne 'hbyyyyincr' && ($i > 0 || $exist_itemnumber));
```

also if we start adding with the SAME barcode again the SAME number of elements (i.e. adding 5 from barcode 12341234 two times) it adds ZERO items second time because all five barcodes occupied and if we have `$i++` incremented in for loop so we just skip all 5 items and done.

And for the question above about barcode generation – barcode generated and incremented so this should be fine.

kohaputti commented:
why should the barcode start from from the last copy added? I don't think that is even possible so why not just remove those two lines and always get an available barcode?

nugged commented:
not last copy added but barcode starts from barcode entered in the form:
![image](https://user-images.githubusercontent.com/295324/66923811-2b7aa600-f032-11e9-900f-5f35d0c3c970.png)
so the rest of my explanation above about that too. 

If I enter that code with $i++ in "for" only, and let's say I'll add 5 items,
then I submit the same form again with same code and:
it loops against already existing items in DB 5 times and sone. 0 items added.

![image](https://user-images.githubusercontent.com/295324/66923996-7d233080-f032-11e9-8606-fd26737ddbde.png)

kohaputti commented:
Ok, so I think this feature is for the case when librarian has e.g 10 barcode stickers printed and waiting to be added to the items. So the barcode stickers might be 1,2,3...,10. Could we here in the item addition loop start from the given barcode and then for the following items find just the next biggest barcode by integer comparison? We are able to do it in constant time and without the loop and it is then always guaranteed to work unless we run out of the barcode space.

nugged commented:
yes it will be jumping over existing and go further, but if we will do as I did first time, removing $i++ from "for" line and leaving $i++ when only real item added,

I'll commit the code in 2-3 hours to show, have to commute now between offices :)...

nugged commented on cataloguing/additem.pl:
```diff
@@ -679,6 +686,7 @@ sub get_common_fields_template {
                     # That way, all items are added, even if there was some already existing barcodes
                     # FIXME : Please note that there is a risk of infinite loop here if we never find a suitable barcode
+                    # NOT FIX but possibility to detect issue: ($total_iterations - $i) > $close_to_infinity counter on the beginning of loop.
                     $i++;
```
@kohaputti this is incremented when real item added, so `i `will to towards `number_of_copies` goal anyway

nugged commented on cataloguing/additem.pl:
```diff
@@ -630,8 +630,15 @@ sub get_common_fields_template {
+            my $total_iterations = 0;
+            my $close_to_infinity = 10_000;
 
+            for (my $i = 0; $i < $number_of_copies; $total_iterations++) {
```
`total_iterations` used to prevent the big difference between `i` for the case of the infinite loop (let's say "unknown border case when barcode not generated 10000 times – but against process hanging in such loop).

nugged commented:
@kohaputti: after discussion in IRC and learning two options of barcode generations:
`$barcodeobj->next_value`
`C4::Barcodes::ValueBuilder::hbyyyyincr::get_barcode`
I can say that we always have the barcode or should generate the error, so I am leaving i++ in for clause as you asked and removed that bottom i++ and added "no_next_barcode" error. And that's now all, we can ask to merge! Here:

nugged commented:
@kohaputti – this one is it.

kohaputti commented:
Hi, please squash/rebase the commits and submit it first to community. Makes review easier.

nugged commented:
squashed to obvious single simple changes, no trash inside.

kohaputti commented on cataloguing/additem.pl:
```diff
@@ -643,8 +643,13 @@ sub get_common_fields_template {
+                    }
+                    else {
+                        push @errors, "no_next_barcode";
+                        $itemrecord = $record;
```
This is undefined after the loop so not necessary

nugged replied:
`$record` undefined? It's defined and used to search next barcode

nugged replied:
ah, it cleans up AFTER loop anyway, yes, I see, so we can't mimic another same way it used when error generated to refill the form with item data

nugged replied:
because I mimicked how this one with `push @errors, "no_next_barcode";` worked above on the line 616-621, because then later in line 992 if we have itemrecord defined it is used to re-fill form, after ERROR case, for user convenience as I understand.

I can still program some extra logic to not to clean up itemrecord on error, better than remove this line, eh, should I?

kohaputti replied:
I don't know how $itemrecord is used, if it is needed in case of error then do the extra logic, if it is not then just remove that line.

kohaputti commented:
Please add KD-4000 to the commit title, and the description:

> we have two options of barcode generations:
$barcodeobj->next_value
C4::Barcodes::ValueBuilder::hbyyyyincr::get_barcode
we always have the barcode or should generate the error.

doesn't make make much sense for me, maybe you mean something like:

> The i++ being in the for clause is sufficient because the $barcodeobj->next_value and C4::Barcodes::ValueBuilder::hbyyyyincr::get_barcode statements return a valid barcode always except when we are out of barcodes. Therefore, we can remove the other i++ statement inside the for loop. In the case where we run out of barcodes this adds the error "no_next_barcode" to the list of errors.

nugged commented:
yes, @kohaputti, I picked comments from various commits and trials... Thanks for fixing the readability. Updated.

So we done with this one, I'll try to propagate it to community version?

kohaputti commented:
@nugged you edited the pull request description instead of commit. Propagate to community any relevant parts of this patch (if there is any).

nugged commented:
Edited now both (https://github.com/nugged/Koha/commit/2bb305d9cdf6036f49fe9f10aab20ec53df75316)
+ preparing patch for the Community version.

kohaputti commented:
Since this appears to fix a bug introduced by KD-543 maybe we could actually put as title "KD-543: (squashable) Add requested amount of items instead of half the amount".
